core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -985,15 +985,23 @@ public void addFailedEngineListener(Engine.FailedEngineListener failedEngineList
     }
 
     public void updateBufferSize(ByteSizeValue shardIndexingBufferSize, ByteSizeValue shardTranslogBufferSize) {
+        Engine engine = engineUnsafe();
+        if (engine == null) {
+            logger.debug("updateBufferSize: engine is closed; skipping");
+            return;
+        }
+
         final EngineConfig config = engineConfig;
         final ByteSizeValue preValue = config.getIndexingBufferSize();
+
         config.setIndexingBufferSize(shardIndexingBufferSize);
+
         // update engine if it is already started.
-        if (preValue.bytes() != shardIndexingBufferSize.bytes() && engineUnsafe() != null) {
-            // its inactive, make sure we do a refresh / full IW flush in this case, since the memory
-            // changes only after a "data" change has happened to the writer
-            // the index writer lazily allocates memory and a refresh will clean it all up.
-            if (shardIndexingBufferSize == EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER && preValue != EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER) {
+        if (preValue.bytes() != shardIndexingBufferSize.bytes()) {
+            if (shardIndexingBufferSize == EngineConfig.INACTIVE_SHARD_INDEXING_BUFFER) {
+                // it's inactive: make sure we do a refresh / full IW flush in this case, since the memory
+                // changes only after a "data" change has happened to the writer
+                // the index writer lazily allocates memory and a refresh will clean it all up.
```

Comment:
>  the previous code first did the settings update on the IW, then hit refresh..

Good catch, I'll move the `onSettingsChange` up before the refresh.

The previous code (and still with this change) only does a refresh when the buffer is changed down to its inactive value (512 KB) ... other buffer changes still won't be "noticed" by IW until another doc or delete op is run against the shard.

>  Also, refresh is optional (maybeRefreshBlocking) - what happens if there is a large buffer but no data to refresh?

I think that's OK: it means IW is not holding onto any memory.  IW still gets the buffering change, and the next op that hits it will use that new buffer size.